crates/linker/src/macho.rs

```diff
@@ -1286,98 +1295,137 @@ fn surgery_macho_help(
+                            _ => internal_error!(
+                                "Invalid relocation for libc symbol, {:+x?}: {name}",
+                                rel
+                            ),
                         }
-                        match rel.1.size() {
-                            32 => {
-                                let data = (target as i32).to_le_bytes();
-                                exec_mmap[base..base + 4].copy_from_slice(&data);
-                            }
-                            64 => {
-                                let data = target.to_le_bytes();
-                                exec_mmap[base..base + 8].copy_from_slice(&data);
-                            }
-                            x => {
-                                internal_error!("Relocation size not yet supported: {}", x);
-                            }
-                        }
-                    } else if matches!(app_obj.symbol_by_index(index), Ok(sym) if ["__divti3", "__udivti3", "___divti3", "___udivti3"].contains(&sym.name().unwrap_or_default()))
-                    {
-                        // Explicitly ignore some symbols that are currently always linked.
                         continue;
                     } else {
                         internal_error!(
                             "Undefined Symbol in relocation, {:+x?}: {:+x?}",
                             rel,
                             app_obj.symbol_by_index(index)
-                        );
+                        )
+                    };
+
+                    let virt_base = section_virtual_offset + rel.0 as usize;
+                    let base = section_offset + rel.0 as usize;
+                    let target: i64 = match rel.1.kind() {
+                        RelocationKind::Relative | RelocationKind::PltRelative => {
+                            target_offset - virt_base as i64 + rel.1.addend()
+                        }
+                        RelocationKind::Absolute => {
+                            target_offset + rel.1.addend()
+                                - subtractor
+                                    .take()
+                                    .map(|index| {
+                                        get_target_offset(
+                                            index,
+                                            &app_obj,
+                                            md,
+                                            &symbol_vaddr_map,
+                                            verbose,
+                                        )
+                                        .unwrap_or(0)
+                                    })
+                                    .unwrap()
+                        }
+                        RelocationKind::MachO { value, relative: _ } => match value {
+                            macho::ARM64_RELOC_SUBTRACTOR => {
```

Comment:
Interesting, never seen a subtractor before.

Reply:
It's a special reloc type that is always followed by (well, applied to) `ARM64_RELOC_UNSIGNED`. If we have two symbols `_foo` and `_bar`, and want to store the distance between the two, we can use this reloc pair like so:

```
_foo: 0x1234
...
_bar: 0x2345
...
_other: _bar - _foo
```

and reloc stream

```
...
SUBTRACTOR(_other, _foo)
UNSIGNED(_other, _bar)
...
```

which yields `_other = _bar - _foo`. Additionally, we ought to check for any addend which by default for `UNSIGNED` reloc type is encoded directly in the relocated symbol's placeholder (i.e., current value of `_other` will be the addend that we apply to the equation above). Note that with no `SUBTRACTOR`, `UNSIGNED` simply becomes an absolute address of some target symbol.